<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.10.0"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>LOOS: Exceptions in LOOS and PyLOOS</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<script type="text/javascript" src="clipboard.js"></script>
<script type="text/javascript" src="cookie.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">LOOS<span id="projectnumber">&#160;4.1.0</span>
   </div>
   <div id="projectbrief">The Lightweight Object Oriented Structural analysis library/toolkit</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.10.0 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="index.html">Lightweight Object-Oriented Structure library (LOOS)</a></li>  </ul>
</div>
</div><!-- top -->
<div><div class="header">
  <div class="headertitle"><div class="title">Exceptions in LOOS and PyLOOS</div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p>As of LOOS 2.2.0, how LOOS uses exceptions has changed significantly. This will probably not affect user code that relies on uncaught exceptions to terminate with an error, with a few exceptions (no pun intended).</p>
<p>In general, exceptions thrown by LOOS will derive from the <a class="el" href="classloos_1_1_l_o_o_s_error.html" title="Generic LOOS exception.">LOOSError</a> class, which in turn derives from std::exception. In cases where it would be more appropriate to return a standard exception, such as std::range_error from the operator[], LOOS will throw that instead.</p>
<p>The exceptions you are most likely to encounter are <a class="el" href="classloos_1_1_file_error.html" title="Errors related to File I/O.">FileError</a> and derived classes. These cover errors while reading and writing the various model and trajectory formats. The <a class="el" href="classloos_1_1_file_open_error.html" title="Error while opening a file.">FileOpenError</a> is associated with a problem opening the file (e.g. missing file or incorrect permissions). The <a class="el" href="classloos_1_1_file_read_error.html" title="Errors that occur while reading a file.">FileReadError</a> covers any number of errors while reading from a file. Since some classes (such as trajectory classes and most model classes) read from the file as part of instantiation, this exception may occur during the "opening" process. The <a class="el" href="classloos_1_1_file_read_error_with_line.html" title="Errors that occur while reading a text file (where lines are tracked)">FileReadErrorWithLine</a> exception comes from classes that track the lines read and will report where in the file the error occured. Finally, the <a class="el" href="classloos_1_1_file_write_error.html" title="Errors while writing to files.">FileWriteError</a> covers any error while writing to a file.</p>
<p>Once an exception has been thrown, expecially for the model and trajectory classes, the internal state of the object that threw is undefined. For example, suppose you are reading from a <a class="el" href="classloos_1_1_d_c_d.html" title="Class for reading DCD files.">DCD</a> and a <a class="el" href="classloos_1_1_file_read_error.html" title="Errors that occur while reading a file.">FileReadError</a> occurs. That <a class="el" href="classloos_1_1_d_c_d.html" title="Class for reading DCD files.">DCD</a> object should not be used again after the exception occured.</p>
<h2><a class="anchor" id="autotoc_md40"></a>
Changes Compared to pre-2.2.0 LOOS</h2>
<p>The primary external change in LOOS-2.2 is that <a class="el" href="namespaceloos.html#a01ad12706a6b9884cc8f3a5677a899e3" title="Applies a string-based selection to an atomic group...">selectAtoms()</a> will no longer throw an exception if no atoms are selected. Instead, it will return an empty <a class="el" href="classloos_1_1_atomic_group.html" title="Class for handling groups of Atoms (pAtoms, actually)">AtomicGroup</a>.</p>
<p>There are two major internal changes to how exceptions are handled compared to earlier versions of LOOS. Previously, the standard exceptions were more liberally used (e.g. std::runtime_error, std::logic_error, etc). Most exceptions that come from LOOS are now derived from <a class="el" href="classloos_1_1_l_o_o_s_error.html" title="Generic LOOS exception.">LOOSError</a>. Related to this, the exceptions that LOOS can throw have been streamlined.</p>
<p>The second major change is that the set of exceptions in LOOS have been flattened and placed in the loos namespace. Some classes defined exceptions that were specific to that class. These have either been flattened (e.g. are now in the loos namespace) or have been replaced with more general LOOS exceptions.</p>
<h2><a class="anchor" id="autotoc_md41"></a>
Translating Exceptions to PyLOOS</h2>
<p>In most cases, Swig will convert the exceptions thrown by LOOS into the Python equivalent, or into the wrapped version of the LOOS exception. You can call the LOOSError::what() function to get the error string. Alternatively, the <b>str</b>() method is defined for all LOOS exceptions that are translated into Python.</p>
<p>The following illustrates catching PyLOOS exceptions... </p><div class="fragment"><div class="line"><span class="keywordflow">try</span>:</div>
<div class="line">     model = createSystem(filename)</div>
<div class="line"><span class="keywordflow">except</span> <a class="code hl_class" href="classloos_1_1_l_o_o_s_error.html">loos.LOOSError</a> <span class="keyword">as</span> e:</div>
<div class="line">     print(<span class="stringliteral">&#39;HELP!  There was a problem reading the model...&#39;</span>)</div>
<div class="line">     print(e.what())</div>
<div class="ttc" id="aclassloos_1_1_l_o_o_s_error_html"><div class="ttname"><a href="classloos_1_1_l_o_o_s_error.html">loos::LOOSError</a></div><div class="ttdoc">Generic LOOS exception.</div><div class="ttdef"><b>Definition</b> exceptions.hpp:40</div></div>
</div><!-- fragment --><p>While we have tried to ensure that exceptions that can be thrown by LOOS will be caught and translated for Python, it's possible that a few have been missed. Any exception not properly caught before it reaches Python will result in the Python interpreter exiting. If you find this happening, please file a bug report or email the developers at loos.maintainer [at] gmail.com </p>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.10.0
</small></address>
</body>
</html>
